File: ui/src/extracardinfo/SynergyTrait.ts
import { Utils } from "../config/Utils"
import { SynergyTrait } from "../generated-src/SynergyTrait"

const allSynergyTraits = Utils.enumValues(SynergyTrait) as SynergyTrait[]
const firstSpecialIndex = allSynergyTraits.indexOf(SynergyTrait.enhanced)
const firstDeprecatedIndex = allSynergyTraits.indexOf(SynergyTrait.elusive)
export const deprecatedTraits = allSynergyTraits.slice(firstDeprecatedIndex, allSynergyTraits.length)
export const specialTraits = allSynergyTraits.slice(firstSpecialIndex, allSynergyTraits.length)

export const validSynergies = (Utils.enumValues(SynergyTrait) as SynergyTrait[])
    .filter(traitValue => !deprecatedTraits.includes(traitValue as SynergyTrait))
export const validTraits = (Utils.enumValues(SynergyTrait) as SynergyTrait[])
    .filter(traitValue => !specialTraits.includes(traitValue as SynergyTrait))

export interface SynTraitDisplayGroup {
    groupName: string
    description?: string
    synergyOnly?: boolean
    traits: SynergyTrait[]
}

export const synergyAndTraitGroups: SynTraitDisplayGroup[] = [
    {
        groupName: "Aember and Keys",
        traits: [SynergyTrait.stealsAmber, SynergyTrait.capturesAmber, SynergyTrait.increasesKeyCost,
            SynergyTrait.scalingAmberControl, SynergyTrait.exalt, SynergyTrait.putsAmberOnTarget,
            SynergyTrait.removesCapturedAmber, SynergyTrait.forgesKeys],
    },
    {
        groupName: "Board Control",
        traits: [SynergyTrait.destroys, SynergyTrait.dealsDamage, SynergyTrait.boardClear,
            SynergyTrait.damagesMultiple, SynergyTrait.exhausts],
    },
    {
        groupName: "Efficiency",
        traits: [SynergyTrait.archives, SynergyTrait.archivesRandom, SynergyTrait.drawsCards,
            SynergyTrait.increasesHandSize, SynergyTrait.playsCards, SynergyTrait.playsFromOtherHouse,
            SynergyTrait.shufflesDiscard, SynergyTrait.mills],
    },
    {
        groupName: "Disruption",
        traits: [SynergyTrait.discardsCards, SynergyTrait.purges, SynergyTrait.chains, SynergyTrait.reduces_R_HandSize,
            SynergyTrait.revealsHand, SynergyTrait.controlsHouseChoice],
    },
    {
        groupName: "Recursion",
        traits: [SynergyTrait.returns_R_ToHand, SynergyTrait.returns_R_FromDiscard, SynergyTrait.shuffles_R_IntoDeck,
            SynergyTrait.replays],
    },
    {
        groupName: "Creatures and Artifacts",
        traits: [SynergyTrait.uses, SynergyTrait.readies, SynergyTrait.causesFighting, SynergyTrait.causesReaping,
            SynergyTrait.goodReap, SynergyTrait.goodAction, SynergyTrait.goodPlay, SynergyTrait.goodFight,
            SynergyTrait.goodDestroyed, SynergyTrait.protectsCreatures, SynergyTrait.increasesCreaturePower,
            SynergyTrait.moves],
    },
    {
        groupName: "Other",
        description: `Use "any" for a wild cart trait that matches anything applicable. Like all artifacts or all ` +
            "creatures with even power.",
        traits: [SynergyTrait.any, SynergyTrait.makesTokens, SynergyTrait.preventsDamage, SynergyTrait.alpha, SynergyTrait.omega,
            SynergyTrait.raisesTide, SynergyTrait.lowersTide],
    },
    {
        groupName: "Enhancements",
        description: `The system does not recognize specific enhancements on cards. "Enhanced" lets you synergize ` +
            "with any applicable enhanced cards. Synergizing with bonus pips will synergize regardless of where " +
            "they land.",
        synergyOnly: true,
        traits: [SynergyTrait.enhanced, SynergyTrait.selfEnhanced, SynergyTrait.bonusAmber, SynergyTrait.bonusDraw,
            SynergyTrait.bonusDamage, SynergyTrait.bonusCapture],
    },
    {
        groupName: "House and Deck Synergies",
        description: "These traits are generated for each house and deck and do not follow the normal rules of " +
            "trait strengths.",
        synergyOnly: true,
        traits: [SynergyTrait.creatureCount, SynergyTrait.tokenCount,
            SynergyTrait.totalArmor, SynergyTrait.totalCreaturePower,
            SynergyTrait.highExpectedAmber, SynergyTrait.lowExpectedAmber, SynergyTrait.hasMars],
    },
]

File: src/main/kotlin/coraythan/keyswap/synergy/SynergyTrait.kt
package coraythan.keyswap.synergy

import coraythan.keyswap.generatets.GenerateTs

enum class TraitStrength(val value: Int) {
    STRONG(4),
    NORMAL(3),
    WEAK(2),
    EXTRA_WEAK(1)
}

@GenerateTs
enum class SynergyTrait {

    // Amber and Keys
    stealsAmber,
    capturesAmber,
    increasesKeyCost,
    scalingAmberControl,
    exalt,
    putsAmberOnTarget,
    forgesKeys,
    forgesKeysWithoutAember,
    preventsForging,

    // Board Control
    destroys,
    dealsDamage,
    boardClear,
    damagesMultiple,
    exhausts,

    // Efficiency
    archives,
    archivesRandom,
    drawsCards,
    increasesHandSize,
    playsCards,
    playsFromOtherHouse,
    shufflesDiscard,
    mills,
    revealsTopDeck,

    // Disruption
    discardsCards,
    purges,
    chains,
    reduces_R_HandSize,
    revealsHand,
    controlsHouseChoice,

    // Recursion
    returns_R_ToHand,
    shuffles_R_IntoDeck,
    replays,
    puts_R_OnBottomOfDeck,

    // Creatures and Artifacts
    uses,
    readies,
    causesFighting,
    causesReaping,
    goodReap,
    goodAction,
    goodPlay,
    goodFight,
    goodDestroyed,
    protectsCreatures,
    increasesCreaturePower,
    moves,

    // Other Traits
    any,
    makesTokens,
    scrapValue,
    preventsDamage,
    alpha,
    omega,
    raisesTide,
    lowersTide,
    replaysSelf,
    dangerousRandomPlay,

    // Automatically applied traits below here

    enhanced,
    selfEnhanced,
    bonusAmber,
    bonusDraw,
    bonusDamage,
    bonusCapture,
    bonusDiscard,
    expectedAember,
    capturedAmber,
    targettedCapturedAmber,

    // Deck traits In general these are 50 to 60 percentile = 0, 60+ = 1, 70+ = 2, 80+ = 3 90+ = 4 +=1
    hasMars,

    creatureCount,
    haunted,
    tokenCount,
    totalCreaturePower,
    totalArmor,

    // Special traits
    card,
    // Automatically applied
    highValue,

    // DEPRECATED

    returns_R_FromDiscard,
    highExpectedAmber, // for house: 7=0, 8=1/4, 9=1/2, 10=3/4, 11=1
    // for deck: 22+=1/4, 24+=1/2, 26=3/4, 27+=1

    lowExpectedAmber, // for house: 7=0, 6=1/4, 5=1/2, 4=3/4, 3=1
    // for deck: 18-=1/4, 17-=1/2, 16-=3/4, 14-=1

    removesCapturedAmber,

    highTotalArmor, // 4=1/4, 5=1/2, 7=3/4, 9=1

    // creature count has out-of-house numbers

    // Deck or House only traits
    highTotalCreaturePower, // for house: 22+=1/4, 24+=1/2, 26+=3/4, 28+=1
    // for deck: 68+=1/4, 73+=1/2, 78+=3/4, 84+=1
    lowTotalCreaturePower, // 60-=1/4, 56-=1/2, 51-=3/4, 46-=1

    highCreatureCount, // for house: =<6=0, 7=1/4, 8=1/2, 9 =3/4, 10,11,12=1
    // for deck: 17+=1/4, 18+=1/2, 19+=3/4, 21+=1

    lowCreatureCount, // for house: =>6=0, 5=1/4, 4=1/2, 3=3/4, 2,1,0=1
    // for deck: 16-=1/4, 15-=1/2, 14-=3/4, 13-=1

    upgradeCount, // for house or deck: 1+ = 1/4 2+ = 1/2 3+ = 3/4 4+ = 1
    highArtifactCount, // 4=0, 5=1/4, 6=1/2, 7=3/4, 8+=1
    lowArtifactCount, // 4=0, 3=1/4, 2=1/2, 1=3/4, 0=1

    elusive,
    skirmish,
    poison,
    deploy,
    ward,
    preventsFighting,
    protectsFromEffects,
    preventsRemoval,
    heals,
    distributableDamage,
    stuns,
    sacrificesCreatures,
    addsPowerTokens,
    discardsFromDeck,
    addsArmor,
    usableArtifact,
    staticArtifact,
    reorderDeck,
    revealsBottomDeck,
    controls,
    inAllHouses,
    regenerates;

    // DEPRECATED END

    companion object {
        fun fromTrait(trait: String): SynergyTrait? {
            return try {
                valueOf(trait.lowercase())
            } catch (e: IllegalArgumentException) {
                null
            }
        }
    }
}
